STM32: Correct I2C master error handling

monkiineko opened the pull request:
## Description
If I2C slave support is included, then the I2C error handler would always reset the I2C address, resulting in incorrectly changing the I2C state to listen for a controller configured as I2C master.  This change conditionalizes the address setting to only occur if the controller was in slave mode when the error occurred.

## Status
READY

## Migrations
NO

## Related PRs
None

## Todos
None

## Deploy notes
None

## Steps to test or reproduce
Create a build for an STM32 based board with DEVICE_I2CSLAVE and DEVICE_I2C_ASYNCH enabled (current default).  Create an I2C master and issue a .transfer() call to a non-existent I2C device address, so that the transfer will be NACKed.  HAL_I2C_ErrorCallback() will be called on the NACK and call i2c_slave_address(), resulting in the state being set to LISTEN.  Subsequent I2C transfer attempts to any address will then immediately fail because i2c_active() returns 1 (true) for any state other than READY.

0xc0170 commented:
cc @bcostm @adustm @LMESTM @jeromecoutant

LMESTM commented on targets/TARGET_STM/i2c_api.c:
```diff
@@ -904,7 +904,8 @@ void HAL_I2C_ErrorCallback(I2C_HandleTypeDef *hi2c){
     /*  restore slave address */
-    i2c_slave_address(obj, 0, address, 0);
+    if(address != 0)
+        i2c_slave_address(obj, 0, address, 0);
```
actually, one small remark, code may need small formatting update : space after if and use of braces
if (address != 0) {
    i2c_slave_address(obj, 0, address, 0);
}

0xc0170 commented:
@monkiineko CLA signed? https://developer.mbed.org/contributor_agreement/

monkiineko commented:
@LMESTM Thanks for the review.  I have updated the formatting as you request (my preferred formatting, actually ; I had copied the format from the slave "if" a few lines above, but I see other "if"s in the file using the improved formatting).  I also noticed that i2c_init() would reset the slave field back to 0, so I added a line just before the slave address restore to set slave field to 1, as it should be.

monkiineko commented:
@0xc0170 Yes, I have previously signed the CLA using my mbed.org account, under the name bscott:
https://developer.mbed.org/users/bscott/

LMESTM commented:
@monkiineko  good catch - thanks for joining your efforts to ours !

0xc0170 commented:
/morph test

bridadan commented:
Timing test failure due to CI load, restarting...

/morph test
